File: LeetCode root folder/Path with Maximum Gold.cpp
// tle
class Solution {
  unordered_set<string> visited;
  int row, col;
  vector<vector<int>> matrix;
public:
  int getMaximumGold(vector<vector<int>>& grid) {
      row = grid.size(), col = grid[0].size();
      matrix = grid;
      int res = 0;
      for (int r = 0; r < row; r++) {
          for (int c = 0; c < col; c++) {
              res = max(res, dfs(r, c));
          }
      }
      return res;
  }

  int dfs(int r, int c) {
      string coor = to_string(r) + ',' + to_string(c);
      if (
          (r < 0 || r >= row) ||
          (c < 0 || c >= col) ||
          visited.find(coor) != visited.end() ||
          matrix[r][c] == 0
      ) return 0;

      visited.insert(coor);
      int res = max({dfs(r+1, c), dfs(r-1, c), dfs(r, c+1), dfs(r, c-1)}) + matrix[r][c];
      visited.erase(coor);
      return res;
  } 
};

File: LeetCode root folder/Path with Maximum Gold.py
# garbage question, the same solution in c++ hit a tle everytime for no reason!:4980
# ms
# Beats
# 10.12%
class Solution:
    def getMaximumGold(self, grid: List[List[int]]) -> int:
        row, col = len(grid), len(grid[0])
        res = 0
        visited = set()

        def dfs(r, c):
            if (
                    (r not in range(row)) or
                    (c not in range(col)) or
                    (r, c) in visited or
                    grid[r][c] == 0
            ):
                return 0

            visited.add((r, c))
            res = max(
                dfs(r + 1, c),
                dfs(r - 1, c),
                dfs(r, c + 1),
                dfs(r, c - 1)
            ) + grid[r][c]
            visited.remove((r, c))
            return res

        for r in range(row):
            for c in range(col):
                res = max(res, dfs(r, c))

        return res

